File: src/bacon.c
#include <math.h>
#include <R.h>
#include <Rmath.h>
#include <float.h>
#include <R_ext/Print.h>

void wsmultinom(double *y, double *w, int n, double *mu, double *sig, double *p,
                double *nj, double *sj, double *sj2){

  double prob[3] = {0.0, 0.0, 0.0};
  double prob_tot = 0.;
  int rN[3];

  for(int j = 0; j < 3; j++)
    nj[j] = sj[j] = sj2[j] = 0.0;

  for(int i = 0; i < n; i++){
    
    prob_tot = 0.;
    for(int j = 0; j < 3; j++) {
      prob[j] = p[j]*dnorm(y[i], mu[j], sig[j], 0);
      prob_tot += prob[j];
    }
	
	//normalized probabilities
    for(int j = 0; j < 3; j++)
      prob[j] = prob[j]/prob_tot;

    rmultinom(w[i], prob, 3, rN);

    for(int j = 0; j < 3; j++) {
      nj[j] += (rN[j] == 0 ? 0.0 : rN[j]);
      sj[j] += (rN[j] == 0 ? 0.0 : rN[j]*y[i]);
      sj2[j] += (rN[j] == 0 ? 0.0 : rN[j]*y[i]*y[i]);
    }
  }
}

void fwrs(double *y, int n, double *mu, double *sig, double *p,
          double *nj, double *sj, double *sj2){

  //  Fast weighted random sampling according to: Efraimidisa and Spirakisb
  //  http://www.sciencedirect.com/science/article/pii/S002001900500298X#

  int l = 0;
  double s = 0.0;
  double maxval = 0.0;
  double val = 0.0;
  int z[n];
  double iweights[3] = {0.0, 0.0, 0.0};

  for(int i = 0; i < n; i++) {
    s = 0.0;
    for(int j = 0; j < 3; j++) {
      iweights[j] = p[j]*dnorm(y[i], mu[j], sig[j], 0);
      s += iweights[j];
    }

    for(int j = 0; j < 3; j++)
      iweights[j] = s/(iweights[j]+DBL_EPSILON);

    l = 0;
    maxval = pow(runif(0.0, 1.0), iweights[0]);
    for(int j = 1; j < 3; j++) {
      val = pow(runif(0.0, 1.0), iweights[j]);
      if(maxval < val) {
        maxval = val;
        l = j;
      }
    }

    z[i] = l;
  }

  for(int j = 0; j < 3; j++) {
    nj[j] = sj[j] = sj2[j] = 0.0;
    for(int i = 0; i < n; i++) {
      nj[j] += (z[i] == j ? 1.0 : 0.0);
      sj[j] += (z[i] == j ? y[i] : 0.0);
      sj2[j] += (z[i] == j ? y[i]*y[i] : 0.0);
    }
  }

}

void bacon(double *y, double *w, double *medyR, double *madyR, int *nR, int *niterR, double *levelR, int *binnedR, int *verboseR,
           double *gibbsmu, double *gibbssig, double *gibbsp,
           double *alpha, double *beta, double *lambda, double *tau, double *gamma){

  /*
    Fit three component normal mixture using Gibbs sampler of niter
    starting values use median and mad of y
    output values are full traces of the mixture parameters: mu, sigma, and proportion
  */

  int n = *nR;
  int niter = *niterR;
  int binned = *binnedR;
  int verbose = *verboseR;
  double medy = *medyR;
  double mady = *madyR;
  double level = *levelR;

  double s = 0.0;
  for(int i = 0; i < n; i++)
    s += w[i];

  double threshold = qnorm(1.0 -level/(2.0*s), 0, 1, 0, 0);

  /* starting values */

  double p[3] = {0.0, 0.0, 0.0};
  double mu[3] = {medy, medy-threshold*mady, medy+threshold*mady};
  double sig[3] = {mady, mady, mady};

  for(int i = 0; i < n; i++) {
    p[1] += (y[i] > medy - threshold*mady) ? w[i]/s : 0.0;
    p[2] += (y[i] < medy + threshold*mady) ? w[i]/s : 0.0;
  }
  p[0] = 1.0 - p[1] - p[2];

  /* working variables */

  double nj[3] = {0.0, 0.0, 0.0};
  double sj[3] = {0.0, 0.0, 0.0};
  double sj2[3] = {0.0, 0.0, 0.0};

  if(verbose) {
    Rprintf("threshold = %1.4f\n", threshold);
    Rprintf("Starting values:\np0 = %1.4f, p1 = %1.4f, p2 = %1.4f\n", p[0], p[1], p[2]);
    Rprintf("mu0 = %1.4f, mu1 = %1.4f, mu2 = %1.4f\n", mu[0], mu[1], mu[2]);
    Rprintf("sigma0 = %1.4f, sigma1 = %1.4f, sigma2 = %1.4f\n", sig[0], sig[1], sig[2]);
  }

  for(int k = 0; k < niter; k++) {

    if(binned)
      wsmultinom(y, w, n, mu, sig, p, nj, sj, sj2);
    else
      fwrs(y, n, mu, sig, p, nj, sj, sj2);

    for(int j = 0; j < 3; j++) {
      gibbsmu[3*k+j] = mu[j] = rnorm((lambda[j]*tau[j] + sj[j])/(nj[j] + tau[j]), sqrt(sig[j]*sig[j]/(tau[j] + nj[j])));
      sj2[j] = sj2[j] + mu[j]*(nj[j]*mu[j] - 2*sj[j]);
    }

    s = 0.0;
    for(int j = 0; j < 3; j++) {
      gibbssig[3*k+j] = sig[j] = sqrt(1.0/rgamma(*alpha + 0.5*(nj[j] + 1.0), 1.0/(*beta + 0.5*tau[j]*(mu[j] - lambda[j])*(mu[j] - lambda[j]) + 0.5*sj2[j])));
      gibbsp[3*k+j] = p[j] = rgamma(nj[j] + gamma[j], 1);
      s += p[j];
    }

    for(int j = 0; j < 3; j++)
      gibbsp[3*k+j] = p[j] = p[j]/s;
  }
}

File: R/bacon.R
.bacon <- function(i, object, niter, nbins, trim, level, verbose, priors){

    ##TODO add some kind of trimming?
    
    tstats <- tstat(object)[,i]
    medy <- median(tstats)
    mady <- mad(tstats)
    binned <- !is.null(nbins)

    if(verbose & binned)
        message("Use multinomial weighted sampling...")
    else if(verbose & !binned)
        message("Use fast random weighted sampling...")
    
    if(!is.null(nbins)){
        ##q <- range(tstats) ##sensitive to outlying values
        q <- quantile(tstats, prob=c(1 - trim, trim))
        tstats <- tstats[tstats > q[1] & tstats < q[2]]
        breaks <- seq(q[1], q[2], length = nbins + 1)
        x <- breaks[-c(nbins+1)] + 0.5*(q[2] - q[1])/(nbins) #identical to h$mids
        h <- hist(tstats, breaks = breaks, plot=FALSE)
        w <- h$counts
        tstats <- h$mids
    }
    else
        w <- 1.0+0*tstats

    output <- .C("bacon",
                 y = as.vector(tstats, mode="double"),
                 w = as.vector(w, mode="double"),
                 medy = as.double(medy),
                 mady = as.double(mady),
                 n = as.integer(length(tstats)),
                 niter = as.integer(niter),
                 level = as.double(level),
                 binned = as.integer(binned),
                 verbose = as.integer(verbose),
                 gibbsmu = vector(3*niter, mode="double"),
                 gibbssig = vector(3*niter, mode="double"),
                 gibbsp = vector(3*niter, mode="double"),
                 alpha = as.double(priors$sigma$alpha),
                 beta = as.double(priors$sigma$beta),
                 lambda = as.vector(priors$mu$lambda, mode="double"),
                 tau = as.vector(priors$mu$tau, mode="double"),
                 gamma = as.vector(priors$epsilon$gamma, mode="double"),
                 PACKAGE="bacon")

    gibbsp <- matrix(output$gibbsp, ncol=3, byrow=TRUE,
                     dimnames=list(NULL, paste0("p.", 0:2)))
    gibbsmu <- matrix(output$gibbsmu, ncol=3, byrow=TRUE,
                      dimnames=list(NULL, paste0("mu.", 0:2)))
    gibbssig <- matrix(output$gibbssig, ncol=3, byrow=TRUE,
                       dimnames=list(NULL, paste0("sigma.", 0:2)))

    return(cbind(gibbsp, gibbsmu, gibbssig))
}

##' Gibbs Sampler Algorithm to fit a three component normal mixture to
##' z-scores
##'
##'
##' @title Gibbs sampler
##' @param teststatistics numeric vector or matrix of test-statistics
##' @param effectsizes numeric vector or matrix of effect-sizes
##' @param standarderrors numeric vector or matrix of standard errors
##' @param niter number of iterations
##' @param nburnin length of the burnin period
##' @param nbins default 1000 else bin test-statistics
##' @param trim default 0.999 trimming test-statistics 
##' @param level significance leve used to determine prop. null for
##'     starting values
##' @param verbose default FALSE
##' @param priors list of parameters of for the prior distributions
##' @return object of class-Bacon
##' @examples
##' ##simulate some test-statistic from a normal mixture
##' ##and run bacon
##' y <- rnormmix(2000, c(0.9, 0, 1, 0, 4, 1))
##' bc <- bacon(y)
##' ##extract all estimated mixture parameters
##' estimates(bc)
##' ##extract inflation
##' inflation(bc)
##' ##extract bias
##' bias(bc)
##'
##' ##extract bias and inflation corrected test-statistics
##' head(tstat(bc))
##'
##' ##inspect the Gibbs Sampling output
##' traces(bc)
##' posteriors(bc)
##' fit(bc)
##'
##' ##simulate multiple sets of test-statistic from a normal mixture
##' ##and run bacon
##' y <- matrix(rnormmix(10*2000, c(0.9, 0, 1, 0, 4, 1)), ncol=10)
##' bc <- bacon(y)
##' ##extract all estimated mixture parameters
##' estimates(bc)
##' ##extract only the inflation
##' inflation(bc)
##' ##extract only the bias
##' bias(bc)
##' ##extract bias and inflation corrected P-values
##' head(pval(bc))
##' ##extract bias and inflation corrected test-statistics
##' head(tstat(bc))
##' @author mvaniterson
##' @references Implementation is based on a version from Zhihui Liu
##'     \url{https://macsphere.mcmaster.ca/handle/11375/9368}
##' @export
##' @importFrom BiocParallel bplapply bpworkers bpparam
##' @useDynLib bacon
bacon <- function(teststatistics=NULL, effectsizes=NULL, standarderrors=NULL,
                  niter=5000L, nburnin = 2000L, nbins=1000, trim =0.999, level=0.05, verbose=FALSE,
                  priors = list(sigma = list(alpha = 1.28,
                                             beta = 0.36), ##original uses 0.36*mad(teststatistics)
                                mu = list(lambda = c(0.0, 3.0, -3.0),
                                          tau = c(1000.0, 100.0, 100.0)),
                                epsilon = list(gamma = c(90.0, 5.0, 5.0)))){

    ##create new Bacon-object
    object <- new("Bacon", teststatistics = teststatistics,
                  effectsizes = effectsizes,
                  standarderrors = standarderrors,
                  niter = niter,
                  nburnin = nburnin,
                  priors = priors)

    ##run the Gibbs Sampler
    if(ncol(tstat(object)) > 1){
        nworkers <- bpworkers(bpparam())
        if(nworkers <= 1) {
            message("Did you registered a biocparallel back-end?\n Continuing serial!")
            for(i in 1:ncol(tstat(object)))
                object@traces[,,i] <- .bacon(i, object, niter, nbins, trim, level, verbose, priors)
        }
        else{
            message("Detected ", nworkers, " workers!\n Running in parallel!")
            ret <- bplapply(1:ncol(tstat(object)), .bacon, object=object,
                            niter=niter, nbins=nbins, trim=trim, level=level, verbose=verbose, priors=priors)
            object@traces <- simplify2array(ret)
        }
    } else
        object@traces[,,1] <- .bacon(1, object, niter, nbins, trim=trim, level, verbose, priors)

    ##summarize traces
    for(i in 1:ncol(tstat(object)))
        object@estimates[i,] <- apply(object@traces[-c(1:object@nburnin),,i], 2, mean)

    return(object)
}
